// mongo-ember/lib/mongo-ember.js
// Generated by CoffeeScript 1.7.1
var fs, js, me, mongo, mongoose, path, settings;

path = require('path');

fs = require('fs');

mongoose = require('mongoose');

mongo = require('./mongo');

js = require('./js');

settings = exports.settings = {};

me = module.exports = {
  getSchema: function(name) {
    return mongo.getSchema(name);
  },
  createModel: function(name, schema) {
    var model;
    model = mongo.createModel(name, schema);
    return js.addModel(model);
  },
  loadModels: function(folder) {
    var load;
    load = function(dir) {
      var file, files, _i, _len, _results;
      files = fs.readdirSync(dir);
      _results = [];
      for (_i = 0, _len = files.length; _i < _len; _i++) {
        file = files[_i];
        _results.push((function(file) {
          var name, obj, schema;
          file = path.join(dir, file);
          if (!fs.statSync(file).isDirectory()) {
            if (path.extname(file) === settings.modelExtension) {
              obj = require(file);
              name = Object.keys(obj)[0];
              schema = obj[name];
              return me.createModel(name, schema);
            }
          } else {
            return load(file);
          }
        })(file));
      }
      return _results;
    };
    return load(folder);
  },
  connect: function(connectionString, cb) {
    mongoose.connect(connectionString, {
      server: {
        socketOptions: {
          keepAlive: 1
        }
      }
    });
    mongoose.connection.on('connected', function() {
      console.log('MongoDB connected');
      if (cb != null) {
        return cb();
      }
    });
    mongoose.connection.on('disconnected', function() {
      return console.log('MongoDB disconnected');
    });
    return mongoose.connection.on('error', function(err) {
      return console.log('MongoDB returned with error ' + err);
    });
  },
  setup: function(options) {
    settings.modelExtension = options.modelExtension || '.js';
    settings.modelsFolder = options.modelsFolder || null;
    settings.namespace = options.namespace || null;
    settings.connectionString = options.connectionString || null;
    settings.primaryKey = options.primaryKey || '_id';
    settings.mePath = options.mePath || '/js/me.js';
    this.loadModels(settings.modelsFolder);
    return this;
  },
  start: function(app, cb) {
    app.use(js.middleware);
    if (settings.connectionString != null) {
      return this.connect(settings.connectionString, cb);
    }
  }
};
